util/include/channels/mpsc_bounded_channel.h

```diff
@@ -0,0 +1,226 @@
+class BoundedMpscSender;
+template <typename Msg>
+class BoundedMpscReceiver;
+
+namespace impl {
+
+// A simple, non-optimized, lock based, bounded MPSC channel.
+//
+// This is an internal type used by the BoundedMpscSender and BoundedMpscReceiver and should not be exposed publicly.
+template <typename Msg>
+class BoundedMpscQueue {
+ public:
+  BoundedMpscQueue(size_t capacity) : capacity_(capacity) {}
+
+ private:
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscSender;
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscReceiver;
+
+  std::optional<size_t> size() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    return buf_.size();
+  }
+
+  std::optional<size_t> capacity() {
+    // No need for a lock, as capacity_ is immutable
+    return capacity_;
+  }
+
+  void addSender() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    num_senders_++;
+  }
+
+  void removeSender() {
+    bool num_senders = 0;
+    {
+      const std::lock_guard<std::mutex> lock(mutex_);
+      num_senders_--;
+      num_senders = num_senders_;
+    }
+    if (num_senders == 0) {
+      // Signal the receiver to throw a NoSendersError.
+      cv_.notify_one();
+    }
+  }
+
+  void invalidateReceiver() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    valid_receiver_ = false;
+  }
+
+  const size_t capacity_;
+  std::queue<Msg> buf_;
+  mutable std::mutex mutex_;
+  std::condition_variable cv_;
+  size_t num_senders_ = 0;
+  bool valid_receiver_ = true;
+};
+
+}  // namespace impl
+
+template <typename Msg>
+class BoundedMpscSender {
+ public:
+  // Try to send a message over a bounded channel.
+  //
+  // Return std::nullopt on sucess.
+  // Return the Msg if the buffer is full
+  //
+  // Throws NoReceiversError if the receiver has been destructed.
+  //
+  // It's important to note that there is a race condition, whereby the receiver can be destructed
+  // immediately after a successful send. There is no absolute guarantee of delivery. The
+  // purpose of returning an error is to prevent endlessly retrying and never noticing that the
+  // channel is effectively destroyed.
+  std::optional<Msg> send(Msg&& msg) {
+    {
+      std::lock_guard<std::mutex> lock(queue_->mutex_);
+      if (!queue_->valid_receiver_) {
+        throw NoReceiversError();
+      }
+      if (queue_->buf_.size() == queue_->capacity_) {
+        return msg;
+      }
+      queue_->buf_.push(std::move(msg));
+    }
+    queue_->cv_.notify_one();
+
+    return std::nullopt;
+  }
+
+  std::optional<size_t> size() { return queue_->size(); }
+  std::optional<size_t> capacity() { return queue_->capacity(); }
+
+  ~BoundedMpscSender() {
+    if (!queue_) return;
+    queue_->removeSender();
+  }
+
+  BoundedMpscSender(const BoundedMpscSender& other) : BoundedMpscSender(other.queue_) {}
+  BoundedMpscSender& operator=(const BoundedMpscSender& other) {
+    *this = BoundedMpscSender(other);
+    return *this;
+  }
```

Comment:
If I am not mistaken, these can be made `= default`. Maybe even omit the `= default` move ones. Please correct me if I miss something.

Reply:
They can be default, but we must be explicit about them due to the custom destructor and rule of 5.

Reply:
Nope, the default constructors do not work. The queue does not appear to get copied correctly in that case.

Reply:
Oh, lord, the non-default dtor... You are right. That's extremely tricky, especially in this case. If you don't mention the copy and move ctors at all, you won't get moves if you have a non-default dtor. But if you make both copy and move `= default` it should have moves as in the existing code. Maybe it is still worth it to make the copy ops `= default` for simplicity. Please correct me if I am wrong.

Reply:
I tried making the copy ops default, but I got a NoSendersError. It appears they don't do what we expect.

Reply:
What I mean is that the default copy and move for the single shared_ptr member should suffice as that's what the custom copy ctor is doing. If not, I must be missing something.

Reply:
Ah, I get it. Yes, just making the copy assignment operator default, but leaving the copy constructor implemented makes it all work..

Reply:
Fixed.

Reply:
I knew I was missing something - the copy ctor is calling the ctor taking the queue and, in effect, calling `addSender()`.  I didn't see that in the beginning. Sorry  for the noise and thanks for addressing it.

Reply:
Having said above, if I don't miss something else (don't count on it :D), then `operator=()` should call `removeSender()` on the sender's current queue and then call `addSender()` on the new one.

Since the only way of creating a sender is to create a pair of a sender and a receiver for a specific channel, we might also consider disabling `operator=()`. I don't have a strong opinion on that, though.